apps/omg/lib/omg/ethereum_client_monitor.ex

```diff
@@ -14,50 +14,133 @@
+
+  The implementation assumes ws subscription to the Ethereum client continues to work indefinitely. We'll see how that works in practice.
   """
   use GenServer
   require Logger
   alias OMG.Eth
+  alias OMG.Eth.Encoding
+  alias OMG.Eth.SubscriptionWorker
 
   @default_interval Application.get_env(:omg, :client_monitor_interval_ms)
   @type t :: %__MODULE__{
           interval: pos_integer(),
           tref: reference() | nil,
           alarm_module: module(),
-          raised: boolean()
+          raised: boolean(),
+          ethereum_height: integer | :error,
+          ws_url: binary() | nil
         }
-  defstruct interval: @default_interval, tref: nil, alarm_module: nil, raised: true
+  defstruct interval: @default_interval,
+            tref: nil,
+            alarm_module: nil,
+            raised: true,
+            ethereum_height: :error,
+            ws_url: nil
+
+  def get_ethereum_height do
+    GenServer.call(__MODULE__, :ethereum_height)
+  end
 
   def start_link(args) do
     GenServer.start_link(__MODULE__, args, name: __MODULE__)
   end
 
-  def init([alarm_module]) do
+  def init([_ | _] = opts) do
+    alarm_module = Keyword.get(opts, :alarm_module)
+    false = Process.flag(:trap_exit, true)
     _ = Logger.info("Starting Ethereum client monitor.")
-    install()
-    state = %__MODULE__{alarm_module: alarm_module}
+    install_alarm_handler()
+    ethereum_height = check()
+
+    state = %__MODULE__{
+      alarm_module: alarm_module,
+      ethereum_height: ethereum_height,
+      ws_url: Keyword.get(opts, :ws_url)
+    }
+
     _ = alarm_module.set({:ethereum_client_connection, Node.self(), __MODULE__})
-    _ = raise_clear(alarm_module, state.raised, check())
-    {:ok, tref} = :timer.send_after(state.interval, :health_check)
-    {:ok, %{state | tref: tref}}
+    _ = raise_clear(alarm_module, state.raised, ethereum_height)
+    {:ok, state, {:continue, :ws_connect}}
   end
 
-  # gen_event
+  # gen_event init
   def init(_args) do
     {:ok, %{}}
   end
 
-  def handle_info(:health_check, state) do
-    _ = raise_clear(state.alarm_module, state.raised, check())
+  def handle_continue(:ws_connect, state) do
+    _ = Logger.debug("Ethereum client monitor starting a WS newHeads subscription.")
+
+    params =
+      case state.ws_url do
+        nil -> [listen_to: "newHeads"]
+        ws_url -> [listen_to: "newHeads", ws_url: ws_url]
+      end
+
+    _ = SubscriptionWorker.start_link(params)
+    _ = raise_clear(state.alarm_module, state.raised, state.ethereum_height)
+    {:noreply, state}
+  rescue
+    _ ->
+      _ = Logger.debug("Ethereum client monitor failed at WS newHeads subscription. Health check in #{state.interval}")
+      {:ok, tref} = :timer.send_after(state.interval, :health_check)
+      _ = raise_clear(state.alarm_module, state.raised, :error)
+      {:noreply, %{state | tref: tref}}
+  end
+
+  def handle_call(:ethereum_height, _from, state) do
```

Comment:
As for the monitor, it has many functions related to non-monitoring

Reply:
everything is related to monitoring :) but because it stores height, it's used for that as well

Reply:
but please do tell, what else is non-monitoring?